cnxepub/formatters.py

```diff
@@ -159,14 +159,16 @@ def _content(self):
             return tree_to_html(
                 model_to_tree(self.model), self.extensions).decode('utf-8')
         elif isinstance(self.model, Document):
-            content = self.model.content
             if self.generate_ids:
                 _html = deepcopy(self.model._xml)
                 self._generate_ids(self.model, _html)
-                content = ''.join(utf8([
-                                   isinstance(node, (type(''), type(b''))) and
-                                   node or etree.tostring(node)
-                                   for node in _html.xpath('node()')]))
+            else:
+                _html = self.model._xml
+
+            content = ''.join(utf8([
```

Comment:
At first glance this kind of makes sense, since `self._generate_ids()` only changes ids and doesn't add or remove tags so whether we generate ids or not, we should have the same code to generate the content.

Then I went back to look at why we have code to strip the root tag, and it's because we [used to create an extra div element](https://github.com/openstax/cnx-epub/commit/bf5fbc41bf8f4a68a73099b06b5f6b6e697a6060#diff-6599de93a091ed0fb6b2c7b619b19c48R142):

```python
            if self.generate_ids:
                div = lxml.html.fragment_fromstring(content, 'div')
                self._generate_ids(self.model, div)
                content = ''.join(utf8([
                    isinstance(node, (type(''), type(b''))) and
                    node or etree.tostring(node)
                    for node in div.xpath('node()')]))
```

At some point [we changed it](https://github.com/openstax/cnx-epub/commit/333141fe48110a123c2f0a84e5d4d647813ab19f#diff-6599de93a091ed0fb6b2c7b619b19c48L144) to not add a div but still remove the root tag:

```diff
@@ -141,12 +143,12 @@ class HTMLFormatter(object):
         elif isinstance(self.model, Document):
             content = self.model.content
             if self.generate_ids:
-                div = lxml.html.fragment_fromstring(content, 'div')
-                self._generate_ids(self.model, div)
+                _html = deepcopy(self.model._xml)
+                self._generate_ids(self.model, _html)
                 content = ''.join(utf8([
-                    isinstance(node, (type(''), type(b''))) and
-                    node or etree.tostring(node)
-                    for node in div.xpath('node()')]))
+                                   isinstance(node, (type(''), type(b''))) and
+                                   node or etree.tostring(node)
+                                   for node in _html.xpath('node()')]))
             return content
 
     @property
```

At this point, we have different output depending on whether `self.generate_ids` is true.  If `self.generate_ids` is true, the root tag is removed, if false, the content is unchanged.

We need to have a look at [self.model._xml](https://github.com/openstax/cnx-epub/blob/326e013975183ef97bb6bea84607aafc510bca97/cnxepub/models.py#L85) to see if we should always remove the root tag:

```python
def content_to_etree(content):
    if not content:  # Allow building empty models
        return etree.XML('<div xmlns="http://www.w3.org/1999/xhtml" />')
    xml_parser = etree.XMLParser(ns_clean=True)
    tree = etree.XML(content, xml_parser)
    # Determine if we've been fed a full XHTML page, with a <body> tag:
    bods = tree.xpath('//*[self::body|self::x:body]',
                      namespaces={'x': 'http://www.w3.org/1999/xhtml'})
    if bods:
        bods[0].tag = '{http://www.w3.org/1999/xhtml}div'
        return bods[0]
    else:  # It parsed, so must have been fed a valid XML tree - return it
        return tree
```

If content was empty, we add a `<div>` tag, if the content has a body tag, we change the body tag to a div, otherwise no change.  So I'm not sure if it's correct to just remove the root tag.

---

The other thing I'm concerned about is... we need to be able to create a `Binder`, use `SingleHTMLFormatter()` to convert it into a single html, and then get the `Binder` back by calling `cnxepub.adapters.adapt_single_html`.  I wrote some code to do this and this is the result:

```
content: <p>hello</p>
document content in binder: <p>hello</p>
adapted content: <div xmlns="http://www.w3.org/1999/xhtml" data-type="page" id="ingress"/>

content: <html><body>hello</body></html>
document content in binder: <html:div xmlns:html="http://www.w3.org/1999/xhtml">hello</html:div>
adapted content: <div xmlns="http://www.w3.org/1999/xhtml" data-type="page" id="ingress"/>

content: <html><body><p>hello</p></body></html>
document content in binder: <html:div xmlns:html="http://www.w3.org/1999/xhtml"><p>hello</p></html:div>
adapted content: <div xmlns="http://www.w3.org/1999/xhtml" data-type="page" id="ingress"><p id="39337">hello</p>
  </div>
```

In the first and second cases, the content is gone.. The third case is may be correct?

This is the code I wrote:

```python
# -*- coding: utf-8 -*-       
                              
import io

from cnxepub.adapters import adapt_single_html
from cnxepub.formatters import SingleHTMLFormatter
from cnxepub.models import Binder, Document


def format_and_adapt(content):
    # build a binder with one page
    base_metadata = {
        'publishers': [],
        'created': '2013/03/19 15:01:16 -0500',
        'revised': '2013/06/18 15:22:55 -0500',
        'authors': [
            {'type': 'cnx-id',
             'name': 'Sponge Bob',
             'id': 'sbob'}],
        'editors': [],
        'copyright_holders': [],
        'illustrators': [],
        'subjects': ['Science and Mathematics'],
        'translators': [],
        'keywords': ['Bob', 'Sponge', 'Rock'],
        'title': "Goofy Goober Rock",
        'license_text': 'CC-By 4.0',
        'license_url': 'http://creativecommons.org/licenses/by/4.0/',
        'summary': "<p>summary</p>",
        'version': 'draft',
    }

    metadata = base_metadata.copy()
    metadata.update({'title': 'Kraken'})
    binder = Binder('Kraken', metadata=metadata)
    metadata = base_metadata.copy()
    metadata.update({'title': 'entrée'})
    binder.append(Document('ingress', io.BytesIO(content),
                  metadata=metadata))

    single_html = str(SingleHTMLFormatter(binder))
    adapted_binder = adapt_single_html(single_html)
    print('content: {}'.format(content))
    print('document content in binder: {}'.format(binder[0].content))
    print('adapted content: {}\n'.format(adapted_binder[0].content))


format_and_adapt(b'<p>hello</p>')
format_and_adapt(b'<html><body>hello</body></html>')
format_and_adapt(b'<html><body><p>hello</p></body></html>')
```

Reply:
Good point, I agree on a general "this is very surprising, what the heck is the model doing" way. Stripping out fragment support like I did was a bit … extreme, it seems.
On the other hand, I'm not convinced any actual production workflow is putting anything like case 1 or 2 into a Document - all the pages are full HTML pages initially created from CNXML, with the `<body>` tag converted from the `<content>` tag of the CNXML. Free text is not allowed as a child of `<content>`, so all text will always be wrapped in at least a `<p>`. I think I can write a quick `xpath` test in the DB to confirm that.

Reply:
The thornier case may be what is baking creating, since that output is adapted from a single html to create pages. Yesterday's code was in fact stripping an extra div from the top - when going to the database (via DocumentFormater)

Reply:
What we can do is to add a `<div>` to `<p>hello</p>` in `content_to_etree()`, that will make everything correct then? :thinking:

Reply:
(Love this test code BTW) I just rolled back to pre-fragment-support removal, and case 2 still deletes the content. Case 1 now passes through. Hmm.

Reply:
I added a DocumentContentFormatter output to your test. It's now even more confusing:
```$ python test.py
content: <p>hello</p>
document content in binder: <p>hello</p>
adapted content: <div xmlns="http://www.w3.org/1999/xhtml" data-type="page" id="ingress"/>
formatted content: <html xmlns="http://www.w3.org/1999/xhtml">
  <body><p>hello</p></body>
</html>


content: <html><body>hello</body></html>
document content in binder: <html:div xmlns:html="http://www.w3.org/1999/xhtml">hello</html:div>
adapted content: <div xmlns="http://www.w3.org/1999/xhtml" data-type="page" id="ingress"/>
formatted content: <html xmlns="http://www.w3.org/1999/xhtml">
  <body><div xmlns:html="http://www.w3.org/1999/xhtml">hello</div></body>
</html>


content: <html><body><p>hello</p></body></html>
document content in binder: <html:div xmlns:html="http://www.w3.org/1999/xhtml"><p>hello</p></html:div>
adapted content: <div xmlns="http://www.w3.org/1999/xhtml" data-type="page" id="ingress"><p id="61569">hello</p>
  </div>
formatted content: <html xmlns="http://www.w3.org/1999/xhtml">
  <body><div xmlns:html="http://www.w3.org/1999/xhtml"><p>hello</p></div></body>
</html>
```

Reply:
The missing content has reappeared, though cases 2 and 3 now have an extra top-level div.

Reply:
I thought about adding a `div` unilaterally, as well - the old code basically parsed the input via HTML parser, in recovery mode, which adds <html><body> to the doc, if it doesn't already have that, then found the contents of the body. The original missing content bug, though, seemed to be coming from that HTML parser (and recovery mode, in particular)

Reply:
This test code should be part of the `adapt_single_html` test suite.  I didn't think to add it when I originally wrote the code. :disappointed:

Reply:
100% agreed